blog.md

```diff
@@ -0,0 +1,258 @@
+```
+ % ./target/release/bench -w seq -r seq
+# Write 64.945474ms     (64.945µs/op)   (15397 op/sec)
+# Read  1.000046609s    (24ns/op)       (41666666 op/sec)
+```
+
+With those two lines the last few months of labor bear fruit. 
+I reimplemented the BBolt database in Rust. 
+It is the culmination of my journey where I learned how to build a database, how to design database architecture in Rust, and how to use API design to remove footguns.
+
+# Impetus
+
+The project started last year when I wanted to learn how ETCD and the Raft algorithm worked which then lead me to learning how the BBolt database worked. 
+Following in the software engineering tradition what started as curiosity turned into an all-consuming case of yak shaving. 
+
+# Goals and Standards
+* Built in stable Rust
+* Library API as close to the original as possible
+* On-file representation exactly matches original code 
+* Manage transaction memory without Rc
+* Unsafe behavior validated by Miri
+
+# BBolt Under a Microscope
+Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets. 
+Buckets may store key/value pairs in bytes and sub-Buckets. 
+Sub-buckets may be stored inline inside a parent bucket or on its own page.  
+
+The data types used to represent the BBolt database are relatively straightforward. 
+
+```rust
+/// The Page ID. Page address = `PgId` * meta.page_size
+#[repr(C)]
+pub struct PgId(u64);
+
+/// The Transaction ID. Monotonic and incremented every commit
+#[repr(C)]
+pub struct TxId(u64);
+
+/// `PageHeader` represents the on-file layout of a page header. 
+/// 
+/// `page` in Go BBolt
+#[repr(C)]
+pub struct PageHeader {
+  /// This Page's ID
+  id: PgId,
+  /// Page's type. Branch(0x01), Leaf(0x02), Meta(0x04), or FreeList(0x10)
+  flags: u16,
+  /// Defines the number of items in the Branch, Leaf, and Freelist pages
+  count: u16,
+  /// How many additional meta.page_size pages are included in this page
+  overflow: u32,
+}
+
+/// `BucketHeader` represents the on-file layout of a bucket header.
```

Comment:
This is a bit better. I would pull this out of a codeblock and do a little more explanation in plain English.

Reply:
These are direct copies of the struct definitions. Would it not be better to leave the descriptions as they are in the code?